File: domains/test/python/src/core/exceptions.py
"""
Custom exceptions for the analytics platform.

This module defines all custom exceptions used throughout the platform
to provide clear error handling and debugging information.
"""

from typing import Any, Dict, Optional


class PlatformError(Exception):
    """Base exception for all platform-specific errors."""
    
    def __init__(self, message: str, details: Optional[Dict[str, Any]] = None):
        """Initialize platform error.
        
        Args:
            message: Error message
            details: Additional error details
        """
        super().__init__(message)
        self.message = message
        self.details = details or {}
    
    def __str__(self) -> str:
        """String representation of the error."""
        if self.details:
            return f"{self.message} - Details: {self.details}"
        return self.message


class ConfigurationError(PlatformError):
    """Raised when configuration is invalid or missing."""
    
    def __init__(self, message: str, config_key: Optional[str] = None):
        """Initialize configuration error.
        
        Args:
            message: Error message
            config_key: Configuration key that caused the error
        """
        details = {"config_key": config_key} if config_key else {}
        super().__init__(message, details)


class ValidationError(PlatformError):
    """Raised when data validation fails."""
    
    def __init__(self, message: str, field: Optional[str] = None, value: Any = None):
        """Initialize validation error.
        
        Args:
            message: Error message
            field: Field that failed validation
            value: Invalid value
        """
        details = {}
        if field:
            details["field"] = field
        if value is not None:
            details["value"] = value
        super().__init__(message, details)


class DataIngestionError(PlatformError):
    """Raised when data ingestion fails."""
    
    def __init__(self, message: str, source: Optional[str] = None, 
                 format: Optional[str] = None):
        """Initialize data ingestion error.
        
        Args:
            message: Error message
            source: Data source that failed
            format: Data format involved
        """
        details = {}
        if source:
            details["source"] = source
        if format:
            details["format"] = format
        super().__init__(message, details)


class ProcessingError(PlatformError):
    """Raised when data processing fails."""
    
    def __init__(self, message: str, processor: Optional[str] = None,
                 stage: Optional[str] = None):
        """Initialize processing error.
        
        Args:
            message: Error message
            processor: Processor that failed
            stage: Processing stage where failure occurred
        """
        details = {}
        if processor:
            details["processor"] = processor
        if stage:
            details["stage"] = stage
        super().__init__(message, details)


class TransformationError(ProcessingError):
    """Raised when data transformation fails."""
    
    def __init__(self, message: str, transformer: Optional[str] = None,
                 input_type: Optional[str] = None, output_type: Optional[str] = None):
        """Initialize transformation error.
        
        Args:
            message: Error message
            transformer: Transformer that failed
            input_type: Expected input type
            output_type: Expected output type
        """
        super().__init__(message, processor=transformer)
        if input_type:
            self.details["input_type"] = input_type
        if output_type:
            self.details["output_type"] = output_type


class PipelineError(PlatformError):
    """Raised when pipeline execution fails."""
    
    def __init__(self, message: str, pipeline_name: Optional[str] = None,
                 step: Optional[int] = None, step_name: Optional[str] = None):
        """Initialize pipeline error.
        
        Args:
            message: Error message
            pipeline_name: Name of the failed pipeline
            step: Step number where failure occurred
            step_name: Name of the failed step
        """
        details = {}
        if pipeline_name:
            details["pipeline"] = pipeline_name
        if step is not None:
            details["step_number"] = step
        if step_name:
            details["step_name"] = step_name
        super().__init__(message, details)


class CacheError(PlatformError):
    """Raised when cache operations fail."""
    
    def __init__(self, message: str, key: Optional[str] = None,
                 operation: Optional[str] = None):
        """Initialize cache error.
        
        Args:
            message: Error message
            key: Cache key involved
            operation: Operation that failed (get/set/delete)
        """
        details = {}
        if key:
            details["key"] = key
        if operation:
            details["operation"] = operation
        super().__init__(message, details)


class VisualizationError(PlatformError):
    """Raised when visualization generation fails."""
    
    def __init__(self, message: str, chart_type: Optional[str] = None,
                 export_format: Optional[str] = None):
        """Initialize visualization error.
        
        Args:
            message: Error message
            chart_type: Type of chart that failed
            export_format: Export format that failed
        """
        details = {}
        if chart_type:
            details["chart_type"] = chart_type
        if export_format:
            details["export_format"] = export_format
        super().__init__(message, details)


class SchemaError(ValidationError):
    """Raised when schema validation or operations fail."""
    
    def __init__(self, message: str, schema_name: Optional[str] = None,
                 expected_type: Optional[str] = None, actual_type: Optional[str] = None):
        """Initialize schema error.
        
        Args:
            message: Error message
            schema_name: Name of the schema
            expected_type: Expected data type
            actual_type: Actual data type received
        """
        super().__init__(message)
        if schema_name:
            self.details["schema"] = schema_name
        if expected_type:
            self.details["expected_type"] = expected_type
        if actual_type:
            self.details["actual_type"] = actual_type


class ConnectionError(DataIngestionError):
    """Raised when connection to data source fails."""
    
    def __init__(self, message: str, host: Optional[str] = None,
                 port: Optional[int] = None, protocol: Optional[str] = None):
        """Initialize connection error.
        
        Args:
            message: Error message
            host: Host that failed to connect
            port: Port number
            protocol: Connection protocol
        """
        super().__init__(message)
        if host:
            self.details["host"] = host
        if port:
            self.details["port"] = port
        if protocol:
            self.details["protocol"] = protocol


class TimeoutError(PlatformError):
    """Raised when an operation times out."""
    
    def __init__(self, message: str, operation: Optional[str] = None,
                 timeout_seconds: Optional[float] = None):
        """Initialize timeout error.
        
        Args:
            message: Error message
            operation: Operation that timed out
            timeout_seconds: Timeout duration in seconds
        """
        details = {}
        if operation:
            details["operation"] = operation
        if timeout_seconds:
            details["timeout_seconds"] = timeout_seconds
        super().__init__(message, details)


class ResourceError(PlatformError):
    """Raised when resource limits are exceeded."""
    
    def __init__(self, message: str, resource_type: Optional[str] = None,
                 limit: Optional[Any] = None, current: Optional[Any] = None):
        """Initialize resource error.
        
        Args:
            message: Error message
            resource_type: Type of resource (memory, disk, etc.)
            limit: Resource limit
            current: Current resource usage
        """
        details = {}
        if resource_type:
            details["resource_type"] = resource_type
        if limit is not None:
            details["limit"] = limit
        if current is not None:
            details["current"] = current
        super().__init__(message, details)


def handle_platform_error(func):
    """Decorator to handle platform errors gracefully.
    
    Args:
        func: Function to decorate
        
    Returns:
        Decorated function
    """
    def wrapper(*args, **kwargs):
        try:
            return func(*args, **kwargs)
        except PlatformError as e:
            # Log the platform error with details
            import logging
            logger = logging.getLogger(func.__module__)
            logger.error(f"Platform error in {func.__name__}: {e}")
            raise
        except Exception as e:
            # Wrap unexpected errors in PlatformError
            import logging
            logger = logging.getLogger(func.__module__)
            logger.error(f"Unexpected error in {func.__name__}: {e}")
            raise PlatformError(f"Unexpected error in {func.__name__}: {str(e)}")
    
    return wrapper

File: domains/test/typescript/src/core/exceptions.ts
// src/core/exceptions.ts
/**
 * Custom exception classes for the analytics platform.
 */

import type { ErrorDetails, Timestamp } from '../types.js';
import { getTimestamp } from './base.js';

/**
 * Base exception class for the platform
 */
export abstract class PlatformException extends Error {
  public readonly code: string;
  public readonly timestamp: Timestamp;
  public readonly details?: unknown;
  public readonly context?: Record<string, unknown>;

  constructor(
    message: string,
    code: string,
    details?: unknown,
    context?: Record<string, unknown>
  ) {
    super(message);
    this.name = this.constructor.name;
    this.code = code;
    this.timestamp = getTimestamp();
    this.details = details;
    this.context = context;

    // Maintains proper stack trace
    if (Error.captureStackTrace) {
      Error.captureStackTrace(this, this.constructor);
    }
  }

  public toJSON(): ErrorDetails {
    return {
      code: this.code,
      message: this.message,
      details: this.details,
      stack: this.stack,
      timestamp: this.timestamp,
    };
  }

  public toString(): string {
    return `${this.name} [${this.code}]: ${this.message}`;
  }
}

/**
 * Validation exception
 */
export class ValidationException extends PlatformException {
  constructor(
    message: string,
    public readonly field?: string,
    public readonly value?: unknown,
    public readonly constraints?: Record<string, unknown>
  ) {
    super(
      message,
      'VALIDATION_ERROR',
      { field, value, constraints },
      { validationType: 'field' }
    );
  }
}

/**
 * Configuration exception
 */
export class ConfigurationException extends PlatformException {
  constructor(
    message: string,
    public readonly configKey?: string,
    public readonly expectedType?: string,
    public readonly actualType?: string
  ) {
    super(
      message,
      'CONFIGURATION_ERROR',
      { configKey, expectedType, actualType },
      { configurationType: 'system' }
    );
  }
}

/**
 * Data ingestion exception
 */
export class DataIngestionException extends PlatformException {
  constructor(
    message: string,
    public readonly source?: string,
    public readonly format?: string,
    public readonly recordIndex?: number
  ) {
    super(
      message,
      'DATA_INGESTION_ERROR',
      { source, format, recordIndex },
      { phase: 'ingestion' }
    );
  }
}

/**
 * Processing exception
 */
export class ProcessingException extends PlatformException {
  constructor(
    message: string,
    public readonly processorName?: string,
    public readonly stage?: string,
    public readonly recordId?: string
  ) {
    super(
      message,
      'PROCESSING_ERROR',
      { processorName, stage, recordId },
      { phase: 'processing' }
    );
  }
}

/**
 * Connection exception
 */
export class ConnectionException extends PlatformException {
  constructor(
    message: string,
    public readonly host?: string,
    public readonly port?: number,
    public readonly protocol?: string,
    public readonly retryCount?: number
  ) {
    super(
      message,
      'CONNECTION_ERROR',
      { host, port, protocol, retryCount },
      { type: 'network' }
    );
  }
}

/**
 * Resource exception
 */
export class ResourceException extends PlatformException {
  constructor(
    message: string,
    public readonly resourceType: string,
    public readonly resourceId?: string,
    public readonly limit?: number,
    public readonly current?: number
  ) {
    super(
      message,
      'RESOURCE_ERROR',
      { resourceType, resourceId, limit, current },
      { type: 'resource' }
    );
  }
}

/**
 * Schema exception
 */
export class SchemaException extends PlatformException {
  constructor(
    message: string,
    public readonly schemaName?: string,
    public readonly fieldErrors?: Record<string, string[]>,
    public readonly missingFields?: string[],
    public readonly extraFields?: string[]
  ) {
    super(
      message,
      'SCHEMA_ERROR',
      { schemaName, fieldErrors, missingFields, extraFields },
      { type: 'schema' }
    );
  }
}

/**
 * Pipeline exception
 */
export class PipelineException extends PlatformException {
  constructor(
    message: string,
    public readonly pipelineName?: string,
    public readonly stageName?: string,
    public readonly stageIndex?: number,
    public readonly failedStep?: string
  ) {
    super(
      message,
      'PIPELINE_ERROR',
      { pipelineName, stageName, stageIndex, failedStep },
      { phase: 'pipeline' }
    );
  }
}

/**
 * Cache exception
 */
export class CacheException extends PlatformException {
  constructor(
    message: string,
    public readonly cacheKey?: string,
    public readonly operation?: 'get' | 'set' | 'delete' | 'clear',
    public readonly cacheSize?: number
  ) {
    super(
      message,
      'CACHE_ERROR',
      { cacheKey, operation, cacheSize },
      { type: 'cache' }
    );
  }
}

/**
 * Visualization exception
 */
export class VisualizationException extends PlatformException {
  constructor(
    message: string,
    public readonly chartType?: string,
    public readonly dataPoints?: number,
    public readonly renderTarget?: string
  ) {
    super(
      message,
      'VISUALIZATION_ERROR',
      { chartType, dataPoints, renderTarget },
      { phase: 'visualization' }
    );
  }
}

/**
 * Authentication exception
 */
export class AuthenticationException extends PlatformException {
  constructor(
    message: string,
    public readonly authMethod?: string,
    public readonly username?: string,
    public readonly realm?: string
  ) {
    super(
      message,
      'AUTHENTICATION_ERROR',
      { authMethod, username, realm },
      { type: 'security' }
    );
  }
}

/**
 * Authorization exception
 */
export class AuthorizationException extends PlatformException {
  constructor(
    message: string,
    public readonly resource?: string,
    public readonly action?: string,
    public readonly requiredPermissions?: string[],
    public readonly actualPermissions?: string[]
  ) {
    super(
      message,
      'AUTHORIZATION_ERROR',
      { resource, action, requiredPermissions, actualPermissions },
      { type: 'security' }
    );
  }
}

/**
 * Timeout exception
 */
export class TimeoutException extends PlatformException {
  constructor(
    message: string,
    public readonly operation?: string,
    public readonly timeoutMs?: number,
    public readonly elapsedMs?: number
  ) {
    super(
      message,
      'TIMEOUT_ERROR',
      { operation, timeoutMs, elapsedMs },
      { type: 'timeout' }
    );
  }
}

/**
 * Not found exception
 */
export class NotFoundException extends PlatformException {
  constructor(
    message: string,
    public readonly resourceType?: string,
    public readonly resourceId?: string,
    public readonly searchCriteria?: Record<string, unknown>
  ) {
    super(
      message,
      'NOT_FOUND_ERROR',
      { resourceType, resourceId, searchCriteria },
      { type: 'notfound' }
    );
  }
}

/**
 * Already exists exception
 */
export class AlreadyExistsException extends PlatformException {
  constructor(
    message: string,
    public readonly resourceType?: string,
    public readonly resourceId?: string,
    public readonly existingResource?: unknown
  ) {
    super(
      message,
      'ALREADY_EXISTS_ERROR',
      { resourceType, resourceId, existingResource },
      { type: 'conflict' }
    );
  }
}

/**
 * Invalid operation exception
 */
export class InvalidOperationException extends PlatformException {
  constructor(
    message: string,
    public readonly operation?: string,
    public readonly currentState?: string,
    public readonly allowedStates?: string[]
  ) {
    super(
      message,
      'INVALID_OPERATION_ERROR',
      { operation, currentState, allowedStates },
      { type: 'operation' }
    );
  }
}

/**
 * Exception handler utility
 */
export class ExceptionHandler {
  private static handlers: Map<string, (error: PlatformException) => void> = new Map();

  public static register(code: string, handler: (error: PlatformException) => void): void {
    this.handlers.set(code, handler);
  }

  public static handle(error: PlatformException): void {
    const handler = this.handlers.get(error.code);
    if (handler) {
      handler(error);
    } else {
      console.error('Unhandled exception:', error.toJSON());
    }
  }

  public static async handleAsync(
    fn: () => Promise<unknown>,
    fallback?: unknown
  ): Promise<unknown> {
    try {
      return await fn();
    } catch (error) {
      if (error instanceof PlatformException) {
        this.handle(error);
      } else {
        console.error('Unexpected error:', error);
      }
      return fallback;
    }
  }

  public static wrap<T extends (...args: any[]) => any>(
    fn: T,
    errorTransform?: (error: unknown) => PlatformException
  ): T {
    return ((...args: Parameters<T>) => {
      try {
        const result = fn(...args);
        if (result instanceof Promise) {
          return result.catch((error) => {
            const platformError = errorTransform
              ? errorTransform(error)
              : error instanceof PlatformException
              ? error
              : new PlatformException(
                  error?.message || 'Unknown error',
                  'UNKNOWN_ERROR',
                  error
                );
            this.handle(platformError);
            throw platformError;
          });
        }
        return result;
      } catch (error) {
        const platformError = errorTransform
          ? errorTransform(error)
          : error instanceof PlatformException
          ? error
          : new PlatformException(
              error?.message || 'Unknown error',
              'UNKNOWN_ERROR',
              error
            );
        this.handle(platformError);
        throw platformError;
      }
    }) as T;
  }
}

/**
 * Type guard for platform exceptions
 */
export function isPlatformException(error: unknown): error is PlatformException {
  return error instanceof PlatformException;
}

/**
 * Type guard for specific exception types
 */
export function isValidationException(error: unknown): error is ValidationException {
  return error instanceof ValidationException;
}

export function isConfigurationException(error: unknown): error is ConfigurationException {
  return error instanceof ConfigurationException;
}

export function isDataIngestionException(error: unknown): error is DataIngestionException {
  return error instanceof DataIngestionException;
}

export function isProcessingException(error: unknown): error is ProcessingException {
  return error instanceof ProcessingException;
}

export function isConnectionException(error: unknown): error is ConnectionException {
  return error instanceof ConnectionException;
}

/**
 * Create typed exception factory
 */
export function createExceptionFactory<T extends PlatformException>(
  ExceptionClass: new (...args: any[]) => T
) {
  return (...args: ConstructorParameters<typeof ExceptionClass>): T => {
    return new ExceptionClass(...args);
  };
}